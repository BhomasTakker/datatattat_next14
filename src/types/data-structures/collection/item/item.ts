import { BaseInfo, Details, Avatar } from "../base";

export type CollectionItem = BaseInfo & {
	details?: Details;
	avatar?: Avatar;
};
