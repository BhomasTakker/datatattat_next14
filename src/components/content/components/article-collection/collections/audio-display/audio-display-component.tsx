"use client";
import { useRef, useState } from "react";
import { AudioPlayer } from "../../../audio-player/audio-player";
import styles from "./audio-display.module.scss";

import { CollectionItem } from "@/types/data-structures/collection/item/item";
import { InViewComponent } from "@/components/ui/in-view/in-view";
import { Interaction } from "../../article/interaction/interactions";
import { InteractionsOptions } from "../../article/interaction/interactions-map";
import { articleMetaLoader, articleRenderer, articleTemplate } from "../utils";
import { WithData } from "@/components/ui/with-data/with-data";
import { DisplayArticle } from "./display-article";
import {
	AudioDisplayOptions,
	AudioVerticalScrollerSize,
} from "./audio-display";
import H5AudioPlayer from "react-h5-audio-player";

type AudioDisplayComponentProps = {
	articles: CollectionItem[];
} & AudioDisplayOptions;

export const AudioDisplayComponent = ({
	articles = [],
	...options
}: AudioDisplayComponentProps) => {
	const firstArticle = articles[0] || {};
	const src = firstArticle.src;
	const { size = AudioVerticalScrollerSize.medium } = options;
	// if no src then we'll need to handle that
	const [activeSrc, setActiveSrc] = useState(src);
	const template = articleTemplate(styles);
	const [item, setItem] = useState(firstArticle);

	const onClickHnd = (item: CollectionItem) => {
		if (!item || !item.src) {
			return;
		}
		// shouldn't have both
		setActiveSrc(item.src);
		setItem(item);
		// we should specify autoplay
	};

	// bit of a hack to get the variant
	const variant = (item?.media?.mediaType as string) || "default";

	return (
		<div className={styles.container}>
			{/* Display article */}
			<DisplayArticle key={item.title} item={item} />
			<div className={styles.audioPlayer} data-testid="audio-player">
				<AudioPlayer src={activeSrc} variant={variant} />
			</div>
			<ul
				className={`${styles.articles} ${styles[size]}`}
				data-testid="articles-list"
			>
				{articles.map((item) => (
					<li key={item.src}>
						<InViewComponent
							options={{
								threshold: 0,
								triggerOnce: true,
							}}
							template={<div className={styles.template} />}
						>
							<Interaction
								key={item.title}
								type={InteractionsOptions.Click}
								onClick={() => onClickHnd(item)}
							>
								{/* Remove this - we wouldn't get meta on these guys
								you have src or you dont
								 */}
								<WithData
									getter={articleMetaLoader(item)}
									callback={articleRenderer(styles)}
									template={template}
								/>
							</Interaction>
						</InViewComponent>
					</li>
				))}
			</ul>
		</div>
	);
};
