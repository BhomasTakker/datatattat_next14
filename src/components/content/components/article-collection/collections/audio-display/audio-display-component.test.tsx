import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { AudioDisplayComponent } from "./audio-display-component";
import { CollectionItem } from "../../../../../../types/data-structures/collection/item/item";

// Mock child components
jest.mock("../../../audio-player/audio-player", () => ({
	AudioPlayer: ({ src }: { src: string }) => (
		<div data-testid="mock-audio-player">{src}</div>
	),
}));
jest.mock("./display-article", () => ({
	DisplayArticle: ({ item }: { item: CollectionItem }) => (
		<div data-testid="display-article">{item?.title}</div>
	),
}));
jest.mock("../../../../../../components/ui/in-view/in-view", () => ({
	InViewComponent: ({ children }: { children: React.ReactNode }) => (
		<>{children}</>
	),
}));
jest.mock("../../article/interaction/interactions", () => ({
	Interaction: ({ children, onClick }: any) => (
		<div data-testid="interaction" onClick={onClick}>
			{children}
		</div>
	),
}));
jest.mock("../../../../../../components/ui/with-data/with-data", () => ({
	WithData: ({ template }: { template: React.ReactNode }) => <>{template}</>,
}));
jest.mock("../utils", () => ({
	articleMetaLoader: () => jest.fn(),
	articleRenderer: () => jest.fn(),
	articleTemplate: () => <div data-testid="article-template" />,
}));

const mockArticles: CollectionItem[] = [
	{ src: "audio1.mp3", title: "Audio 1" } as CollectionItem,
	{ src: "audio2.mp3", title: "Audio 2" } as CollectionItem,
];

describe("AudioDisplayComponent", () => {
	it("renders without crashing and displays the first article", () => {
		render(<AudioDisplayComponent articles={mockArticles} />);
		expect(screen.getByTestId("display-article")).toHaveTextContent("Audio 1");
		expect(screen.getByTestId("mock-audio-player")).toHaveTextContent(
			"audio1.mp3"
		);
	});

	it("renders a list of articles", () => {
		const { container } = render(
			<AudioDisplayComponent articles={mockArticles} />
		);
		const interactions = screen.getAllByTestId("interaction");
		expect(interactions.length).toBe(2);
		expect(container).toMatchSnapshot();
	});

	it("updates displayed article and audio player src on article click", () => {
		render(<AudioDisplayComponent articles={mockArticles} />);
		const interactions = screen.getAllByTestId("interaction");
		// Click the second article
		fireEvent.click(interactions[1]);
		// The displayed article should update
		waitFor(() => {
			expect(screen.getByTestId("display-article")).toHaveTextContent(
				"Audio 2"
			);
		});

		expect(screen.getByTestId("mock-audio-player")).toHaveTextContent(
			"audio2.mp3"
		);
	});

	it("handles empty articles array gracefully", () => {
		render(<AudioDisplayComponent articles={[]} />);
		expect(screen.getByTestId("display-article")).toBeInTheDocument();
		expect(screen.getByTestId("mock-audio-player")).toHaveTextContent("");
	});

	it("applies the correct size class", () => {
		render(<AudioDisplayComponent articles={mockArticles} size="medium" />);
		const ul = screen.getByRole("list");
		expect(ul.className).toMatch(/medium/);
	});
});
