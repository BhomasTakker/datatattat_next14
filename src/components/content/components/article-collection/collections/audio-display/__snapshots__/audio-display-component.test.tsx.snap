// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioDisplayComponent renders a list of articles 1`] = `
<div>
  <div
    class="container"
  >
    <div
      data-testid="display-article"
    >
      Audio 1
    </div>
    <div
      class="audioPlayer"
      data-testid="audio-player"
    >
      <div
        data-testid="mock-audio-player"
      >
        audio1.mp3
      </div>
    </div>
    <ul
      class="articles medium"
      data-testid="articles-list"
    >
      <li>
        <div
          data-testid="interaction"
        >
          <div
            data-testid="article-template"
          />
        </div>
      </li>
      <li>
        <div
          data-testid="interaction"
        >
          <div
            data-testid="article-template"
          />
        </div>
      </li>
    </ul>
  </div>
</div>
`;
